### DISCUSSION TITLE
grpc-netty-shaded for ManagedChannelBuilder in open telemetry

### DISCUSSION THREAD

**Discussion by hkdev27** [open-telemetry/opentelemetry-java#3680] [Q&A]
While running jaeger implementation for java always getting the following error message 
"main" java.lang.NoSuchMethodError: 'void com.google.common.base.Preconditions.checkArgument(boolean, java.lang.String, char, java.lang.Object)'

Is this due to lib version mismatch. 
I tried with grpc-netty-shaded 1.29.0.

**Comment by anuraaga** [comment]
@hkdev27 Yes this is a version issue of Guava, a requirement for gRPC - try to force a newer version in your build. If other conflicts prevent you from doing this, then you'll need to use the `exporter-otlp-http-trace` exporter and the HTTP receiver on the collector instead of gRPC to avoid the Guava dependency.

**Reply by hkdev27** [reply]
As per requirements I must use jaeger exporter only. Autoconfigure really works fine with jaeger even today. I'm wondering how, the under line implementation works in that case.

**Reply by anuraaga** [reply]
Ah sorry yeah didn't catch that this is Jaeger - the HTTP exporter for that is the `jaeger-thrift` exporter which may help.

> Autoconfigure really works fine with jaeger 

Do you mean that you don't have problems when using autoconfigure, only when manually initializing the SDK? Autoconfigure doesn't really do anything that different I think - perhaps you coincidentally are getting a newer Guava version in that case based on the POM dependency tree.

**Reply by hkdev27** [reply]
@anuraaga I am new to this technology so I understand my question might sound vague. So I am attaching the code here if you can point out the gap it would be very helpful.

At this point I can bring up my application but I don't see service getting registered with jaeger. 

public class OpenTelemetryConfiguration {

	static OpenTelemetry initOpenTelemetry(String jaegerHost, int jaegerPort) {

		Resource serviceNameResource =
				Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "otel-jaeger-example"));

		// Create a channel towards Jaeger end point    
		ManagedChannel jaegerChannel = ManagedChannelBuilder.forAddress(jaegerHost, jaegerPort).usePlaintext().build();

		// Export traces to Jaeger
		
		JaegerGrpcSpanExporter jaegerExporter =
				JaegerGrpcSpanExporter.builder()
				.setEndpoint("http://localhost:14250")
				.setChannel(jaegerChannel)
				.setTimeout(30, TimeUnit.SECONDS)
				.build();

		SdkTracerProvider tracerProvider =
				SdkTracerProvider.builder()
				.addSpanProcessor(SimpleSpanProcessor.create(jaegerExporter))
				.setResource(Resource.getDefault().merge(serviceNameResource))
				.build();

		OpenTelemetrySdk openTelemetry =
				OpenTelemetrySdk.builder().setTracerProvider(tracerProvider).build();

		// it's always a good idea to shut down the SDK cleanly at JVM exit.
		Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::close));

		return openTelemetry;
	}

}

Main class:

@SpringBootApplication
@OpenAPIDefinition
public class AddressApplication {

	public static void main(String[] args) {

		
		  OpenTelemetry openTelemetry =
		  OpenTelemetryConfiguration.initOpenTelemetry("localhost", 14250);
		 

		SpringApplication.run(AddressApplication.class, args);
	}

}

And POM looks like: 

<dependency>
	        <groupId>io.opentelemetry</groupId>
	        <artifactId>opentelemetry-api</artifactId>
	        <version>1.6.0</version>
	      </dependency>
	     <dependency>
		      <groupId>io.opentelemetry</groupId>
		      <artifactId>opentelemetry-sdk</artifactId>
		       <version>1.6.0</version>
		 </dependency>
		 <dependency>
		      <groupId>io.opentelemetry</groupId>
		      <artifactId>opentelemetry-semconv</artifactId>
		       <version>1.6.0-alpha</version>
		 </dependency>
		 <dependency>
		      <groupId>io.opentelemetry</groupId>
		      <artifactId>opentelemetry-exporter-jaeger</artifactId>
		       <version>1.6.0</version>
		 </dependency>
		<dependency>
		      <groupId>io.grpc</groupId>
		      <artifactId>grpc-protobuf</artifactId>
		       <version>1.41.0</version>
		 </dependency>
		 <dependency>
		      <groupId>io.grpc</groupId>
		      <artifactId>grpc-netty</artifactId>
		       <version>1.41.0</version>
		 </dependency>

If I replace <artifactId>grpc-netty</artifactId> with <artifactId>grpc-netty-shaded</artifactId> then I'm getting the earlier mentioned issue.

**Reply by anuraaga** [reply]
What happens if you use dependency management with `opentelemetry-bom` and `grpc-bom` instead? Both projects will not behave well if any artifacts happen to not be aligned properly so it's a possible cause. If you still have issues and are able to post your full POM, that would be great.

**Reply by hkdev27** [reply]
Following is the full POM.

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.4.11</version>
		<relativePath/> 
	</parent>
	<groupId>com.example.address</groupId>
	<artifactId>address</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<name>address</name>
	<description>User Address Storage Service</description>
	<properties>
		<java.version>16</java.version>
	</properties>
	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
		    <groupId>org.springdoc</groupId>
		    <artifactId>springdoc-openapi-ui</artifactId>
		    <version>1.5.10</version>
		</dependency>
		
		<dependency>
	        <groupId>io.opentelemetry</groupId>
	        <artifactId>opentelemetry-api</artifactId>
	        <version>1.6.0</version>
	      </dependency>
	     <dependency>
		      <groupId>io.opentelemetry</groupId>
		      <artifactId>opentelemetry-sdk</artifactId>
		       <version>1.6.0</version>
		 </dependency>
		 <dependency>
		      <groupId>io.opentelemetry</groupId>
		      <artifactId>opentelemetry-semconv</artifactId>
		       <version>1.6.0-alpha</version>
		 </dependency>
		 <dependency>
		      <groupId>io.opentelemetry</groupId>
		      <artifactId>opentelemetry-exporter-jaeger</artifactId>
		       <version>1.6.0</version>
		 </dependency>
		<dependency>
		      <groupId>io.grpc</groupId>
		      <artifactId>grpc-protobuf</artifactId>
		       <version>1.41.0</version>
		 </dependency>
		 <dependency>
		      <groupId>io.grpc</groupId>
		      <artifactId>grpc-netty</artifactId>
		       <version>1.41.0</version>
		 </dependency>
		
		<dependency>
		    <groupId>org.springframework.boot</groupId>
		    <artifactId>spring-boot-starter-data-jpa</artifactId>
		</dependency>
		<dependency>
			<groupId>com.h2database</groupId>
			<artifactId>h2</artifactId>
			<scope>runtime</scope>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>

</project>

**Reply by anuraaga** [reply]
By the way you mentioned `I tried with grpc-netty-shaded 1.29.0.` - did you confirm you have the same problem when everything is set to 1.41.0 (and please try the boms to confirm)?

**Reply by hkdev27** [reply]
Yes, I'm running into same problem for 1.41.0 as with 1.29.0

**Reply by hkdev27** [reply]
I'm doing the POC so if you want I can provide the sample project.

**Reply by anuraaga** [reply]
If you can provide the code that would be awesome!

**Reply by hkdev27** [reply]
You can find the code here..
https://github.com/hkdev27/opentelemetry-jaeger-integration

**Reply by hkdev27** [reply]
@anuraaga Any update or pointers for me ?

**Reply by anuraaga** [reply]
Hi @hkdev27 - sorry just took a look at your repo. The code looks like what I'd expect to work, and when I run `AddressApplicationTests` in your repo, the test passes. Is there another way to reproduce the exception message you are seeing?

**Reply by hkdev27** [reply]
@anuraaga You can try and execute it as spring boot application. Also after execution did get the telemetry data ?

**Reply by anuraaga** [reply]
I played with your code and found that there doesn't seem to be a classpath issue anymore so you were almost there. The last issue was using the global OpenTelemetry - you should inject the SDK you initialized instead.

https://github.com/hkdev27/opentelemetry-jaeger-integration/compare/master...anuraaga:fixes?expand=1

With this change, I could confirm the span gets sent to Jaeger

**Reply by hkdev27** [reply]
@anuraaga Thanks a ton, you are really awesome.

**Reply by hkdev27** [reply]
Hi @anuraaga I need one more help from you. I was trying service to service communication. How can I get extract the trace context at service when the call originates from another service. I looked around the solution it seems I need to inject and extract context but I'm using feign client so I'm able to set the inject using propagators. Can you provide some insights.
